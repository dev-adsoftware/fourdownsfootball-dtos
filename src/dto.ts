import { ClassConstructor, plainToClass } from 'class-transformer';
import { validateSync, ValidationError } from 'class-validator';

export abstract class Dto {
  public init(obj: Record<string | number, unknown>): this {
    Object.assign(
      this,
      plainToClass(this.constructor as ClassConstructor<this>, { ...obj }),
    );
    return this.validate(true);
  }

  public serialize(): string {
    return JSON.stringify(this);
  }

  private validate(whitelist: boolean): this {
    const errors = validateSync(this, {
      whitelist,
      validationError: { target: false },
    });
    if (errors.length > 0) {
      throw new Error(
        `${this.constructor.name}: ${this.validationErrors(errors)}`,
      );
    }

    return this;
  }

  private validationErrors(errors: ValidationError[]): string {
    return errors
      .map((e) => {
        if (e.constraints) return Object.values(e.constraints).join(', ');
        return String(e);
      })
      .join(', ');
  }
}
