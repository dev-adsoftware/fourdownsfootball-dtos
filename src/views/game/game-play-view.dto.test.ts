import { GamePlayView } from './game-play-view.dto';

describe('given: missing data', () => {
  describe('when: I initialize', () => {
    it('then: error matched snapshot', async () => {
      try {
        expect(new GamePlayView().init({})).not.toBeDefined();
      } catch (e) {
        expect(e).toMatchSnapshot();
      }
    });
  });
});

describe('given: missing nested data', () => {
  describe('when: I initialize', () => {
    it('then: error matched snapshot', async () => {
      try {
        expect(
          new GamePlayView().init({
            offenseFormation: {},
            defenseFormation: {},
          }),
        ).not.toBeDefined();
      } catch (e) {
        expect(e).toMatchSnapshot();
      }
    });
  });
});
