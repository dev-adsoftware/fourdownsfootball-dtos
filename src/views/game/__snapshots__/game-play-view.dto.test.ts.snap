// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`given: missing data when: I initialize then: error matched snapshot 1`] = `[Error: GamePlayView: id must be a string, offenseSequence must be a number string, offensePlay must be a string, offenseFormation must be an object, each value in offensePlayers must be a string, offensePlayers must contain at least 11 elements, offensePlayers must contain not more than 11 elements]`;

exports[`given: missing nested data when: I initialize then: error matched snapshot 1`] = `
[Error: GamePlayView: id must be a string, offenseSequence must be a number string, offensePlay must be a string, An instance of an object has failed the validation:
 - property offenseFormation.name has failed the following constraints: isString 
 - property offenseFormation.sequence has failed the following constraints: isNumberString 
 - property offenseFormation.formationType has failed the following constraints: isEnum 
 - property offenseFormation.positionMap has failed the following constraints: isString, arrayMaxSize, arrayMinSize 
, each value in offensePlayers must be a string, offensePlayers must contain at least 11 elements, offensePlayers must contain not more than 11 elements, An instance of an object has failed the validation:
 - property defenseFormation.name has failed the following constraints: isString 
 - property defenseFormation.sequence has failed the following constraints: isNumberString 
 - property defenseFormation.formationType has failed the following constraints: isEnum 
 - property defenseFormation.positionMap has failed the following constraints: isString, arrayMaxSize, arrayMinSize 
]
`;
